File: frontend/lib/spotify.ts
import SpotifyWebApi from 'spotify-web-api-node';

const scopes = [
  'user-read-private',
  'user-read-email',
  'user-top-read',
  'user-read-recently-played',
].join(' ');

const params = {
  scope: scopes,
};

const queryParamString = new URLSearchParams(params);

const spotifyApi = new SpotifyWebApi({
  clientId: process.env.SPOTIFY_CLIENT_ID,
  clientSecret: process.env.SPOTIFY_CLIENT_SECRET,
  redirectUri: process.env.SPOTIFY_REDIRECT_URI,
});

export const LOGIN_URL = `https://accounts.spotify.com/authorize?${queryParamString.toString()}&client_id=${process.env.SPOTIFY_CLIENT_ID}&response_type=code&redirect_uri=${encodeURIComponent(process.env.SPOTIFY_REDIRECT_URI || '')}`;

export { spotifyApi };

File: backend/utils/spotify.go
package utils

import (
	"net/http"
	"os"

	"golang.org/x/oauth2"
)

// SpotifyConfig holds the OAuth2 configuration for Spotify
var SpotifyOauthConfig = &oauth2.Config{
	ClientID:     os.Getenv("SPOTIFY_CLIENT_ID"),
	ClientSecret: os.Getenv("SPOTIFY_CLIENT_SECRET"),
	RedirectURL:  os.Getenv("SPOTIFY_REDIRECT_URI"),
	Scopes: []string{
		"user-read-private",
		"user-read-email",
		"user-top-read",
		"user-read-recently-played",
		"user-read-currently-played",
	},
	Endpoint: oauth2.Endpoint{
		AuthURL:  "https://accounts.spotify.com/authorize",
		TokenURL: "https://accounts.spotify.com/api/token",
	},
}

// GetSpotifyClient creates an HTTP client with the access token
func GetSpotifyClient(token string) *http.Client {
	return &http.Client{
		Transport: &oauth2.Transport{
			Source: oauth2.StaticTokenSource(
				&oauth2.Token{AccessToken: token},
			),
		},
	}
}
